crates/iceberg/src/expr/predicate.rs

```diff
@@ -172,6 +223,146 @@ pub enum Predicate {
     Set(SetExpression<Reference>),
 }
 
+impl Bind for Predicate {
+    type Bound = BoundPredicate;
+
+    fn bind(self, schema: SchemaRef, case_sensitive: bool) -> Result<BoundPredicate> {
+        match self {
+            Predicate::And(expr) => {
+                let bound_expr = expr.bind(schema, case_sensitive)?;
+
+                let [left, right] = bound_expr.inputs;
+                Ok(match (left, right) {
+                    (_, r) if matches!(&*r, &BoundPredicate::AlwaysFalse) => {
+                        BoundPredicate::AlwaysFalse
+                    }
+                    (l, _) if matches!(&*l, &BoundPredicate::AlwaysFalse) => {
+                        BoundPredicate::AlwaysFalse
+                    }
+                    (left, r) if matches!(&*r, &BoundPredicate::AlwaysTrue) => *left,
+                    (l, right) if matches!(&*l, &BoundPredicate::AlwaysTrue) => *right,
+                    (left, right) => BoundPredicate::And(LogicalExpression::new([left, right])),
+                })
+            }
+            Predicate::Not(expr) => {
+                let bound_expr = expr.bind(schema, case_sensitive)?;
+                let [inner] = bound_expr.inputs;
+                Ok(match inner {
+                    e if matches!(&*e, &BoundPredicate::AlwaysTrue) => BoundPredicate::AlwaysFalse,
+                    e if matches!(&*e, &BoundPredicate::AlwaysFalse) => BoundPredicate::AlwaysTrue,
+                    e => BoundPredicate::Not(LogicalExpression::new([e])),
+                })
+            }
+            Predicate::Or(expr) => {
+                let bound_expr = expr.bind(schema, case_sensitive)?;
+                let [left, right] = bound_expr.inputs;
+                Ok(match (left, right) {
+                    (_, r) if matches!(&*r, &BoundPredicate::AlwaysTrue) => {
+                        BoundPredicate::AlwaysTrue
+                    }
+                    (l, _) if matches!(&*l, &BoundPredicate::AlwaysTrue) => {
+                        BoundPredicate::AlwaysTrue
+                    }
```

Comment:
Out of curiosity, why not collapse this into a single statement?
```suggestion
                    (l, r) if matches!(&*r, &BoundPredicate::AlwaysTrue) || matches!(&*l, &BoundPredicate::AlwaysTrue) => {
                        BoundPredicate::AlwaysTrue
                    }
```